README.md
# gatsby-transformer-cloudinary

Provides three ways to use [Cloudinary](https://cloudinary.com) with Gatsby: 1) Upload images in `File` nodes to Cloudinary. 2) Upload remote images by their URL to Cloudinary. 3) Create nodes for images that have already been uploaded to Cloudinary.

Each of the three methods above create `CloudinaryAsset` nodes compatible with [`gatsby-image`](https://www.gatsbyjs.org/packages/gatsby-image/).

You’ll need a [Cloudinary account](https://cloudinary.com) to use this plugin. They have a generous free tier, so for most of us this will stay free for quite a while.

[Live demo](https://gatsby-transformer-cloudinary.netlify.com/) ([source](https://github.com/jlengstorf/gatsby-transformer-cloudinary))

> **DISCLAIMER:** If you try running this demo's source code on your own computer, you might face issues as the demo uses assets and [transformations](https://cloudinary.com/documentation/chained_and_named_transformations#named_transformations) from the author’s Cloudinary account. Before running, please remove them or replace them with images and transformations from your own Cloudinary account.

## Features

- Upload local project media assets to a secure remote CDN
- Upload remote media assets to a secure remote CDN
- Utilize media assets on Cloudinary in gatsby-image
- Use gatsby-image `fluid` and `fixed` formats on Cloudinary assets
- Retrieve media files in optimized formats with responsive breakpoints
- Utilize all Cloudinary transformations including chained transformations in gatsby's data layer

## Example usage

Here's the plugin in action to fetch a fixed asset using the `useStaticQuery` API of gatsby:

```jsx harmony
import React from 'react';
import { graphql, useStaticQuery } from 'gatsby';
import Image from 'gatsby-image';

export default () => {
  const data = useStaticQuery(graphql`
    query {
      file(name: { eq: "marketplace-banner" }) {
        childCloudinaryAsset {
          fixed {
            ...CloudinaryAssetFixed
          }
        }
      }
    }
  `);

  return (
    <div>
      <h2>Here goes something</h2>
      <Image fixed={data.file.childCloudinaryAsset.fixed} alt="banner" />
    </div>
  );
};
```

## Installation

This transformer automatically creates childCloudinaryAsset nodes for `File` nodes created by [`gatsby-source-filesystem`](https://www.gatsbyjs.org/packages/gatsby-source-filesystem/).

This transformer also allows you to pass URLs directly to Cloudinary to side-step the need to first download files to your development machine. This can be achieved by calling the `createRemoteImageNode` function from an `onCreateNode` function.

Install the plugins using either `npm` or `yarn`.

```sh
npm install --save gatsby-transformer-cloudinary gatsby-source-filesystem
```

```sh
yarn add gatsby-transformer-cloudinary gatsby-source-filesystem
```

## How to use

### Set up environment variables

Add the data that shouldn’t be committed to Git into `.env.development`:

```sh
# Find these values at https://cloudinary.com/console/
CLOUDINARY_CLOUD_NAME=<your cloud name>
CLOUDINARY_API_KEY=<your API key>
CLOUDINARY_API_SECRET=<your API secret>
```

> **NOTE:** you’ll also need to set these environment variables in your build system (i.e. Netlify).

### Configure the plugin

In your `gatsby-config.js`, point `gatsby-source-filesystem` to images in your app, then set up `gatsby-transformer-cloudinary` with your credentials.

```js
// Load the environment variables.
require('dotenv').config({
  path: `.env.${process.env.NODE_ENV}`,
});

module.exports = {
  plugins: [
    {
      resolve: `gatsby-source-filesystem`,
      options: {
        name: `images`,
        path: `${__dirname}/src/images`,
      },
    },
    {
      resolve: 'gatsby-transformer-cloudinary',
      options: {
        cloudName: process.env.CLOUDINARY_CLOUD_NAME,
        apiKey: process.env.CLOUDINARY_API_KEY,
        apiSecret: process.env.CLOUDINARY_API_SECRET,
        uploadFolder: 'gatsby-cloudinary',
      },
    },
  ],
};
```

### Upload remote images

To directly upload images to Cloudinary from remote sources, you can use the `createRemoteImageNode` function:

```js
// gatsby-node.js
import { createRemoteImageNode } from 'gatsby-transformer-cloudinary';

// This example assumes "post" nodes are created in a `sourceNodes` function.
const POST_NODE_TYPE = 'post';

export async function onCreateNode({
  node,
  actions: { createNode },
  createNodeId,
  createContentDigest,
  reporter,
}) {
  // In this example, "post" nodes sometimes have a "cover_photo_url" that's a link to an image.
  if (node.internal.type === POST_NODE_TYPE && node.cover_photo_url) {
    await createRemoteImageNode({
      url: node.cover_photo_url,
      parentNode: node,
      relationshipName: 'coverPhoto',
      createNode,
      createNodeId,
      createContentDigest,
      reporter,
    });
  }
}
```

### Use images already on Cloudinary

To create GraphQL nodes for images that are already uploaded to Cloudinary, you need to create nodes containing data that describe the images on Cloudinary. For example, you might have a `post` node that has a cover photo stored on Cloudinary. The data in the post node should look something like...

```js
{
  title: "How to beat the pandemic blues",
  publishedAt: "2020-07-26T21:55:13.358Z",
  coverPhoto: {
    cloudinaryAssetData: true,
    cloudName: "my-amazing-blog",
    publicId: "blue-blue-blue",
    originalHeight: 360,
    originalWidth: 820,
    defaultBase64: "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mMMXG/8HwAEwAI0Bj1bnwAAAABJRU5ErkJggg==",
    defaultTracedSVG: "data:image/svg+xml,%3Csvg%20height%3D%229999%22%20viewBox%3D%220%200%209999%209999%22%20width%3D%229999%22%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%3E%3Cpath%20d%3D%22m0%200h9999v9999h-9999z%22%20fill%3D%22%23f9fafb%22%2F%3E%3C%2Fsvg%3E",
  }
}
```

The `coverPhoto` property in the node above will be deleted and replaced by `gatsby-transformer-cloudinary` with a `CloudinaryAsset` node that can be used with [`gatsby-image`](https://www.gatsbyjs.org/packages/gatsby-image/). This transformation will be done for any top-level properties of nodes that have `cloudinaryAssetData: true`, and values `cloudName`, `publicId`, `originalHeight`, and `originalWidth` properties. The top-level property name, `coverPhoto` in the example above, will be the name of the relationship between the parent node and the `CloudinaryAsset` node that will be created.

The property `defaultBase64` in the node above can be used by your CMS/backend API to provide precomputed or cached base64 URIs for your images. The provided string must comply with [RFC 2397](https://tools.ietf.org/html/rfc2397). This base64 image will be used unless `ignoreDefaultBase64: true` is set in your GraphQL query. In cases where you prefer to have an accurate base64 image with the same transformations applied as you full-size image, you should use `ignoreDefaultBase64: true` in your GraphQL query. When a defaultBase64 property is not supplied or `ignoreDefaultBase64` is true, an API call to Cloudinary will be made when resolving your GraphQL queries to fetch the base64 image.

When providing `defaultBase64` properties, it's recommended that you set the plugin option `alwaysUseDefaultBase64` to true in development. This may result in your base64 images looking different in development and production, but it will also result in much faster development build times as fewer API calls to Cloudinary will be made. The `alwaysUseDefaultBase64` plugin option overrides the `ignoreDefaultBase64` GraphQL query parameter and forces `gatsby-transformer-cloudinary` to always use `defaultBase64` images when they are provided.

The property `defaultTracedSVG` in the node above can be used by your CMS/backend to provide precomputed or cached SVG placeholders for your images. The provided string must comply with [RFC 2397](https://tools.ietf.org/html/rfc2397). It should also be encoded with something like JavaScript's `encodeURIComponent()`.

No API calls to Cloudinary for base64 images will be made if your GraphQL queries do not request base64 images.

### Plugin options

In `gatsby-config.js` the plugin accepts the following options:

| option                         | type      | required | default value | description                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 |
| ------------------------------ | --------- | -------- | ------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `cloudName`                    | `String`  | false    | n/a           | Cloud name of your Cloudinary account, can be obtained from your [Cloudinary console](https://cloudinary.com/console/). This should be stored and retrieved as an environment variable.                                                                                                                                                                                                                                                                                                     |
| `apiKey`                       | `String`  | false    | n/a           | API Key of your Cloudinary account, can be obtained from your [Cloudinary console](https://cloudinary.com/console/). This should be stored and retrieved as an environment variable.                                                                                                                                                                                                                                                                                                        |
| `apiSecret`                    | `String`  | false    | n/a           | API Secret of your Cloudinary account, can be obtained from your [Cloudinary console](https://cloudinary.com/console/). This should be stored and retrieved as an environment variable.                                                                                                                                                                                                                                                                                                     |
| `uploadFolder`                 | `String`  | false    | n/a           | An optional folder name where the uploaded assets will be stored on Cloudinary.                                                                                                                                                                                                                                                                                                                                                                                                             |
| `fluidMaxWidth`                | `Int`     | false    | 1000          | The maximum width needed for an image. If specifying a width bigger than the original image, the width of the original image is used instead. Used when calculating breakpoints.                                                                                                                                                                                                                                                                                                            |
| `fluidMinWidth`                | `Int`     | false    | 200           | The minimum width needed for an image. Used when calculating breakpoints.                                                                                                                                                                                                                                                                                                                                                                                                                   |
| `createDerived`                | `Boolean` | false    | true          | If `true`, create and keep the derived images of the selected breakpoints during the API call. If false, images generated during the analysis process are thrown away. This option is ignored if `useCloudinaryBreakpoints` is `false`. It's recommended that you enable `createDerived` if `useCloudinaryBreakpoints` is true to store the breakpoint images and prevent them from being recalculated on every build.                                                                      |
| `breakpointsMaxImages`         | `Integer` | false    | 5             | Set maximum number of responsive breakpoint images generated and returned on image upload. If `useCloudinaryBreakpoints` is false, then exactly `breakpointsMaxImages` breakpoints will be created.                                                                                                                                                                                                                                                                                         |
| `useCloudinaryBreakpoints`     | `Boolean` | false    | false         | If `true`, then Cloudinary will be requested to automatically find the best breakpoints for each image. It's recommended that this option be set to `false` in development because this option uses one Cloudinary transformation for every image uploaded to Cloudinary plus one transformation for every derived image created while calculating breakpoints.                                                                                                                             |
| `overwriteExisting`            | `Boolean` | false    | false         | Whether to overwrite existing assets with the same public ID. When set to false, return immediately if an asset with the same Public ID was found. It's recommended that this is set to false in development as each image overwrite costs one Cloudinary transformation.                                                                                                                                                                                                                   |
| `enableDefaultTransformations` | `Boolean` | false    | false         | `true` will add the `q_auto` and `f_auto` transformations to images for quality and format optimizations.                                                                                                                                                                                                                                                                                                                                                                                   |
| `alwaysUseDefaultBase64`       | `Boolean` | false    | false         | When `alwaysUseDefaultBase64` is true, `gatsby-transformer-cloudinary` will always use `defaultBase64` images when they are provided to the GraphQL layer. It's recommended that you set `alwaysUseDefaultBase64` to true in your development environment and provide `defaultBase64` properties for any images already uploaded to Cloudinary. Doing so will result in faster and cheaper builds because no Cloudinary API calls will need to be made when resolving your GraphQL queries. |

The options `cloudName`, `apiKey`, and `apiSecret` are required if any images will be uploaded to Cloudinary during the build process. If you're solely using images already uploaded to Cloudinary, then these options can be safely omitted.

> Note: Each derived image created for a breakpoint will consume one Cloudinary transformation. Enable the `useCloudinaryBreakpoints` option with care. If the `createDerived` option is enabled, transformations will only be consumed when the images are first created. However, created images will consume Cloudinary storage space. If `overwriteExisting` is enabled, each image that you upload will consume one transformation each time your Gatsby cache gets cleared and the image gets re-uploaded. For this reason, it's recommended that you keep `overWriteExisting` disabled and instead set the `overwriteExisting` parameter of `createRemoteImageNode` on a per-image basis when you know that an image has actually been updated.

### Query for images

Assuming you have an image called “avatar.jpg” in your `src/images/` directory, you can use it in a component like this:

```jsx harmony
import React from 'react';
import { graphql, useStaticQuery } from 'gatsby';
import Image from 'gatsby-image';

export default () => {
  const data = useStaticQuery(graphql`
    query {
      file(name: { eq: "avatar" }) {
        childCloudinaryAsset {
          fluid {
            ...CloudinaryAssetFluid
          }
        }
      }
    }
  `);

  return <Image fluid={data.file.childCloudinaryAsset.fluid} alt="avatar" />;
};
```

### Fragments

The fragments below can be used when querying your Cloudinary assets:

- `CloudinaryAssetFluid`
- `CloudinaryAssetFluid_noBase64`
- `CloudinaryAssetFluid_tracedSVG`
- `CloudinaryAssetFixed`
- `CloudinaryAssetFixed_noBase64`
- `CloudinaryAssetFixed_tracedSVG`

### Avoiding stretched images using the fluid type

As mentioned previously, images using the fluid type are stretched to match the container’s width and height. In the case where the image’s width or height is smaller than the available viewport, the image will stretch to match the container, potentially leading to unwanted problems and worsened image quality.

The `CloudinaryAssetFluidLimitPresentationSize` fragment can be used to to `gatsby-image` not to stretch an image larger than its maximum dimensions regardless of the size of its container:

```graphql
query {
  file(name: { eq: "avatar" }) {
    childCloudinaryAsset {
      fluid {
        ...CloudinaryAssetFluid
        ...CloudinaryAssetFluidLimitPresentationSize
      }
    }
  }
}
```

## Manual Usage

It’s also possible to manually create `gatsby-image`-friendly `fixed` and `fluid` objects by importing helper functions from the transformer.

This is an advanced use case — if possible, try not to do this when Gatsby’s data layer is an option. This is intended for cases where assets are already on Cloudinary and moving them to the Gatsby project would be too time-intensive to be reasonable.

> **NOTE:** This approach is async, which means you’ll end up with content jumping unless you manually account for the image area. You’ve been warned.

### Manually creating fluid images

```js
import React from 'react';
import Image from 'gatsby-image';
import { getFluidImageObject } from 'gatsby-transformer-cloudinary';

export default () => {
  const [fluid, setFluid] = useState(false);

  useEffect(() => {
    getFluidImageObject({
      public_id: 'gatsby-cloudinary/jason',
      cloudName: 'jlengstorf',
      originalHeight: 3024,
      originalWidth: 4032,
      breakpoints: [200, 400, 600, 800],
      transformations: ['ar_16:10', 'c_fill'],
      chained: ['e_grayscale,e_tint:100:663399:0p:white:100p', 't_lwj'],
    }).then(result => setFluid(result));
  }, []);

  return fluid ? <Image fluid={fluid} alt="Jason" /> : <p>loading...</p>;
};
```

### Manually creating fixed images

```js
import React from 'react';
import Image from 'gatsby-image';
import { getFixedImageObject } from 'gatsby-transformer-cloudinary';

export default () => {
  const [fixed, setFixed] = useState(false);

  useEffect(() => {
    getFixedImageObject({
      public_id: 'gatsby-cloudinary/jason',
      cloudName: 'jlengstorf',
      originalHeight: 3024,
      originalWidth: 4032,
    }).then(result => setFixed(result));
  }, []);

  return fixed ? <Image fixed={fixed} alt="Jason" /> : <p>loading...</p>;
};
```

## API

This plugin can support both the `fixed` and `fluid` formats for `gatsby-image`.

Both `fixed` and `fluid` accept arguments. All arguments are optional.

### Arguments for both `fixed` and `fluid`

| argument              | type        | required | default                | description                                                                                                                                                                                                                                                                                                                                                                      |
| --------------------- | ----------- | -------- | ---------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `cloudName`           | `String`    | true     | `n/a`                  | Cloud name of your Cloudinary account, can be obtained from your [Cloudinary console](https://cloudinary.com/console/). This should be stored and retrieved as an environment variable.                                                                                                                                                                                          |
| `public_id`           | `String`    | true     | `n/a`                  | Public ID of the image to retrieve from Cloudinary. This can be obtained from your Cloudinary account.                                                                                                                                                                                                                                                                           |
| `transformations`     | `[String!]` | false    | `[]`                   | Array of transformations to be applied to the image.                                                                                                                                                                                                                                                                                                                             |
| `chained`             | `[String!]` | false    | `[]`                   | An array of chained transformations to be applied to the image.                                                                                                                                                                                                                                                                                                                  |
| `defaults`            | `[String!]` | false    | `["f_auto", "q_auto"]` | Default transformation applied to the image                                                                                                                                                                                                                                                                                                                                      |
| `originalHeight`      | `Int`       | true     | `n/a`                  | Height of the image fetched. This is required in gatsby-image to calculate the aspect ratio of the image.                                                                                                                                                                                                                                                                        |
| `originalWidth`       | `Int`       | true     | `n/a`                  | Desired width of the image. This is required in gatsby-image to calculate the aspect ratio.                                                                                                                                                                                                                                                                                      |
| `base64Width`         | `String`    | false    | `30`                   | Base64 width of the image.                                                                                                                                                                                                                                                                                                                                                       |
| `version`             | `Boolean`   | false    | `false`                | Version number of image if applicable, eg. 300124291, 1241983.                                                                                                                                                                                                                                                                                                                   |
| `ignoreDefaultBase64` | `Boolean`   | false    | false                  | If this parameter is set to true, then an API call will be made to Cloudinary when resolving your GraphQL queries to fetch a base64 image regardless of whether a `defaultBase64` image was provided for the image. This parameter can be overridden by the `alwaysUseDefaultBase64` plugin option in environments where build speed and economy of Cloudinary usage is desired. |

### Arguments for `fixed`

| argument | type  | default | description                                  |
| -------- | ----- | ------- | -------------------------------------------- |
| `height` | `Int` | `n/a`   | The height that the image should display at. |
| `width`  | `Int` | `400`   | The width that the image should display at.  |

### Arguments for `fluid`


CHANGELOG.md
# Version Next

Improvements:

- Only throw an error on missing Cloudinary credentials if those credentials are actually needed to upload an image to Cloudinary.

# Version 2.1.1

Additions:

- defaultTracedSVG values are now passed along as tracedSVG values.

Improvements:

- Improved base64 caching so that if a second request for the same base64 image is made before the first response is received, only one request is made.

Fixes:

- base64 images are no longer generated unless a query requesting them is run.

# Version 2.1.1

Additions:

- Added logging for each time we have to fetch a base64 image from Cloudinary to explain long query steps in the Gatsby build process.

Fixes:

- Fluid images use defaultBase64 images when they are provided.

# Version 2.1.0

Additions:

- Added the ability to use both width and height parameters simultaneously for fixed queries.
- Added the ability to use precomputed base64 images. When precomputed base64 images are used, build times should improve and Cloudinary usage should decrease.

Fixes:

- Deeply nested asset data is now transformed into CloudinaryAsset nodes.

# Version 2.0.0

Changes:

- This is a major version bump to call attention to the change in default behavior introduced in version 1.1.1. (`f_auto` and `q_auto` are no longer added to image URLs by default.)

Fixes:

- Images uploaded using the `createRemoteImageNode` function respect the `overwriteExisting` argument when provided and fall back to using the `overwriteExisting` plugin option.

# Version 1.1.3

Fixes:

- Typo fix.

# Version 1.1.2

Fixes:

- Local images uploaded to Cloudinary now respect the `overwriteExisting` plugin option.

# Version 1.1.1

Changes:

- Added `enableDefaultTransformations` plugin option. When set to true, `f_auto` and `q_auto` are added to all source URLs automatically. Previously, this was on by default. This behavior is now opt-in.

# Version 1.1.0

Additions:

- Added ability to use existing Cloudinary images by marking nodes with `cloudinaryAssetData: true` and providing `cloudName`, `publicId`, `originalHeight`, and `originalWidth` properties.
- Added an optional `height` argument to `fixed` queries.

Improvements:

- Cache base64 images when running queries to prevent duplicate network requests.

Fixes:

- Changed the public_id to be the relative path of files without the extension instead of just the file's name. This fixes an [issue with childrenCloudinaryAsset nodes](https://github.com/cloudinary-devs/gatsby-transformer-cloudinary/issues/42) being created instead of childCloudinaryAsset nodes.

# Version 1.0.1

Additions:

- Added CloudinaryAssetFluidLimitPresentationSize fragment.
- Added presentationHeight and presentationWidth to CloudinaryAssetFluid.

# Version 1.0.0

Breaking changes:

- The default for `fluidMinWidth` has been decreased from 200 to 50 to match Cloudinary defaults.
- The default for `createDerived` has been changed to false.
- The default for `breakpointsMaxImages` has been increased from 5 to 20 to match Cloudinary defaults.
- Breakpoint calculations are no longer requested by default when uploading an image. (This is controlled by the new `useCloudinaryBreakpoints` option.)
- Image uploads no longer overwrite images with the same public ID by default. (This is controlled by the new `overwriteExisting` option.)

Other changes:

- Added `createRemoteImageNode` to allow uploading remote images directly to Cloudinary without downloading them locally first.
- Added Jest tests.
- Added the ability to calculate breakpoints locally to avoid consuming Cloudinary transformation credits while developing.
- Increased the timeout when uploading images to Cloudinary from 60 seconds to 5 minutes.

# Version 0.3.5

- Beginning of changelog.


get-image-objects.js
const axios = require('axios');
const { getPluginOptions } = require('./options');
// todo: for fluid images when original width and height is not set, use width and height of image as is and make full width
// todo: investigate graphQL query overrides for width and height and abstract control to user through gql query

// Define default width values for fluid, fixed and base64 images
const DEFAULT_BASE64_WIDTH = 30;
const DEFAULT_FIXED_WIDTH = 400;

const base64Cache = {};

// Create Cloudinary image URL with transformations.
const getImageURL = ({
  public_id,
  cloudName,
  transformations = [],
  chained = [],
  defaults = ['f_auto', 'q_auto'],
  version = false,
}) => {
  const baseURL = 'https://res.cloudinary.com/';
  const allTransformations = [transformations.concat(defaults).join()]
    .concat(chained)
    .join('/');

  const imagePath = [
    cloudName,
    '/image/upload/',
    allTransformations,
    version ? `/v${version}/` : '/',
    public_id,
  ]
    .join('')
    .replace('//', '/');

  return baseURL + imagePath;
};

// Fetch and return Base64 image
const getBase64 = async url => {
  if (!base64Cache[url]) {
    const result = await axios.get(url, { responseType: 'arraybuffer' });
    const data = Buffer.from(result.data).toString('base64');
    base64Cache[url] = `data:image/jpeg;base64,${data}`;
  }

  return base64Cache[url];
};

// retrieve aspect ratio if in transformation else create aspect ratio values
const getAspectRatio = (transformations, originalAspectRatio) => {
  const arTransform = transformations.find(t => t.startsWith('ar_'));
  if (!arTransform) {
    return originalAspectRatio;
  }

  const newAspectRatio = arTransform.replace('ar_', '');
  if (newAspectRatio.indexOf(':') === -1) {
    return Number(newAspectRatio);
  }

  const [w, h] = newAspectRatio.split(':').map(Number);

  return w / h;
};

// Create shared image data for both fixed and fluid. Returns src and Base64
const getSharedImageData = async ({
  public_id,
  version,
  cloudName,
  base64Transformations,
  transformations,
  base64Width,
  chained,
}) => {
  const b64Transformations = base64Transformations || transformations;
  const base64URL = getImageURL({
    transformations: b64Transformations.concat(`w_${base64Width}`),
    public_id,
    version,
    cloudName,
    chained,
  });
  const base64 = await getBase64(base64URL);

  const src = getImageURL({
    public_id,
    version,
    cloudName,
    transformations,
    chained,
  });

  return { base64, src };
};

exports.getFixedImageObject = async ({
  public_id,
  cloudName,
  originalHeight,
  originalWidth,
  version = false,
  height,
  width,
  base64Width = DEFAULT_BASE64_WIDTH,
  base64Transformations = [],
  transformations = [],
  chained = [],
}) => {
  const { base64, src } = await getSharedImageData({
    public_id,
    version,
    cloudName,
    base64Transformations,
    transformations,
    base64Width,
    chained,
  });

  const aspectRatio = getAspectRatio(
    transformations,
    originalWidth / originalHeight,
  );

  let displayWidth;
  if (!!width) {
    displayWidth = width;
  } else if (!!height) {
    displayWidth = height * aspectRatio;
  } else if (!height && !width) {
    displayWidth = DEFAULT_FIXED_WIDTH;
  }

  const sizes = [1, 1.5, 2, 3].map(size => ({
    resolution: size,
    width: Math.round(displayWidth * size),
  }));

  const srcSet = sizes
    .filter(size => size.width <= originalWidth)
    .map(size => {
      // Get URL for each image including user-defined transformations.
      const url = getImageURL({
        // Add the size at the end to override width for srcSet support.
        transformations: transformations.concat(`w_${size.width}`),
        chained,
        public_id,
        version,
        cloudName,
      });

      return `${url} ${size.resolution}x`;
    })
    .join();

  return {
    base64,
    height: Math.round(displayWidth / aspectRatio),
    src,
    srcSet,
    width: Math.round(displayWidth),
  };
};

function onlyUnique(element, index, array) {
  return array.indexOf(element) === index;
}

exports.getFluidImageObject = async ({
  public_id,
  cloudName,
  originalWidth,
  originalHeight,
  breakpoints = [200, 400, 600],
  version = false,
  maxWidth,
  base64Width = DEFAULT_BASE64_WIDTH,
  base64Transformations = [],
  transformations = [],
  chained = [],
}) => {
  const aspectRatio = getAspectRatio(
    transformations,
    originalWidth / originalHeight,
  );
  const { fluidMaxWidth } = getPluginOptions();
  const max = Math.min(maxWidth ? maxWidth : fluidMaxWidth, originalWidth);
  const sizes = `(max-width: ${max}px) 100vw, ${max}px`;
  const { base64, src } = await getSharedImageData({
    public_id,
    version,
    cloudName,
    breakpoints,
    base64Transformations,
    transformations,
    base64Width,
    chained,
  });

  const breakpointWidths = breakpoints
    .concat(max) // make sure we get the max size
    .filter(w => w <= max) // don’t add larger sizes
    .sort((a, b) => a - b) // sort in ascending order
    .filter(onlyUnique); // remove duplicates

  const srcSet = breakpointWidths
    .map(breakpointWidth => {
      // Get URL for each image including user-defined transformations.
      const url = getImageURL({
        // Add the size at the end to override width for srcSet support.
        transformations: transformations.concat(`w_${breakpointWidth}`),
        chained,
        public_id,
        version,
        cloudName,
      });

      return `${url} ${breakpointWidth}w`;
    })
    .join();

  const presentationWidth = max;
  const presentationHeight = Math.round(
    (presentationWidth * originalHeight) / originalWidth,
  );

  return {
    aspectRatio,
    base64,
    presentationWidth,
    presentationHeight,
    sizes,
    src,
    srcSet,
  };
};
